components/ai_chat/renderer/page_text_distilling.cc

```diff
@@ -204,4 +230,89 @@ void DistillPageText(
   std::move(callback).Run(contents_text);
 }
 
+void DistillPageTextViaSiteScript(
+    content::RenderFrame* render_frame,
+    std::string script_content,
+    int32_t world_id,
+    base::OnceCallback<void(const std::optional<std::string>&)> callback) {
+  // Concatenates our injector-script (retrieved from the resource bundle) with
+  // our extractor script. Because the injector-script is wrapped in an IIFE, we
+  // leverage custom events dispatched against the window object to communicate
+  // between the two scripts.
+  //
+  // The extractor script is currently hard-coded to request content at level 3.
+  // Level 3 is the "FULL" level, which is the highest content level. The level
+  // could be tied to the currently-active model (e.g., smaller models could be
+  // defaulted to a smaller level).
+  std::string script =
+      script_content + absl::StrFormat(
+                           R"(
+      new Promise(function(resolve, reject) {
+          window.addEventListener('%s', (event) => {
+              if (event?.detail?.result) {
+                  return resolve(event.detail.result);
+              }
+              reject("No result in %s event");
+          });
+      });
+
+      window.dispatchEvent(new CustomEvent(
+          '%s', { detail: { level: 3 } }
+      ));
+    )",
+                           "LEO_DISTILL_RESULT", "LEO_DISTILL_RESULT",
+                           "LEO_DISTILL_REQUESTED");
+
+  blink::WebScriptSource source =
+      blink::WebScriptSource(blink::WebString::FromUTF8(script));
+
+  auto on_script_executed =
+      [](base::OnceCallback<void(const std::optional<std::string>&)> callback,
+         std::optional<base::Value> value, base::TimeTicks start_time) {
+        if (value && value->is_string() && !value->GetString().empty()) {
+          std::move(callback).Run(value->GetString());
+          return;
+        }
+
+        std::move(callback).Run({});
+      };
+
+  // Execute the combined script as a single source
+  render_frame->GetWebFrame()->RequestExecuteScript(
+      world_id, base::span_from_ref(source),
+      blink::mojom::UserActivationOption::kDoNotActivate,
+      blink::mojom::EvaluationTiming::kAsynchronous,
+      blink::mojom::LoadEventBlockingOption::kDoNotBlock,
+      base::BindOnce(on_script_executed, std::move(callback)),
+      blink::BackForwardCacheAware::kAllow,
+      blink::mojom::WantResultOption::kWantResult,
+      // Because we are using a promise to resolve the result, we will use the
+      // `kAwait` option to ensure the promise is resolved before the callback
+      // is invoked.
+      blink::mojom::PromiseResultOption::kAwait);
+}
+
+std::optional<std::pair<std::string, bool>> LoadSiteScriptForHost(
+    std::string_view host) {
+  static const std::map<std::string, std::pair<int, bool>>
```

Comment:
Only objects with a trivial destructor can be declared as `static`, otherwise we can have entanglements between static objects during termination [1].

This should be something like:

```cxx
static constexpr auto kHostToScriptResource = 
           base::MakeFixedFlatMap<std::string_view, std::pair<int, bool>>({...});
```

[1] https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

Reply:
Also, would you mind clarifying what the boolean is for? I ask because looking in the code it looks like this could be a more descriptive enum type.

Reply:
the `bool` is just whether it runs in the main world or not. I don't think we should spend too much time worrying about this per my original comment because this stuff needs to move to a crx anyway so won't be stored like this

Reply:
other than the `static` issue, that is